### PR TITLE
dapp: Add pagination, number of entries dropdown and filtering to table

### PR DESCRIPTION
**Pull request by Norbo11**
The available contracts table is now much more sophisticated and allows for pagination, filtering and choosing the number of entries to display (using a DataTables plugin), all on the front-end ofc.

### CODE REVIEW

File: dapp/src/client/scripts/app.jsx [hunk 1]
```diff
@@ -125,17 +125,13 @@ export class App extends React.Component {
   getAvailableContracts() {
     this.getContracts("/contract/available_contracts")
       .then(data => {
-        const contracts = {}
```
Comment (Norbo11) [inline dapp/src/client/scripts/app.jsx]: We no longer store keys: contracts here, but merely an array of contracts instead

File: dapp/src/client/scripts/app.jsx [hunk 2]
```diff
@@ -152,9 +148,16 @@ export class App extends React.Component {
 
   buyEnergy(contractAddress) {
     const contract = makeContract(contractAddress);
+    var index = 0;
+    for (var i = 0; i < this.state.contracts.length; i++) {
```
Comment (Norbo11) [inline dapp/src/client/scripts/app.jsx]: The above means we have to calculate the index into this array when we want to update the tx field

File: dapp/src/client/scripts/availableContractsTable.jsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  componentWillUnmount() {
+    $('.data-table-wrapper')
+    .find('table')
+    .DataTable()
+    .destroy(true)
+  }
+
+  shouldComponentUpdate(nextProps) {
```
Comment (Norbo11) [inline dapp/src/client/scripts/availableContractsTable.jsx]: If props ever change, we manually clear and re-draw the table

File: dapp/src/client/scripts/availableContractsTable.jsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    const table = $('.data-table-wrapper')
+    .find('table')
+    .DataTable();
+    table.clear();
+    console.log(nextProps.contracts)
+    table.rows.add(nextProps.contracts);
+    table.draw();
+    return false;
```
Comment (Norbo11) [inline dapp/src/client/scripts/availableContractsTable.jsx]: Gotta force react to never re-render this component, as it should be controlled by the JQuery DataTables plugin: https://reactjs.org/docs/integrating-with-other-libraries.html

File: dapp/src/client/scripts/app.jsx (hunk 1 above)
Comment (fangyi-zhou) [reply dapp/src/client/scripts/app.jsx]: This breaks update...

**Pull request merged**